core/jobs/batch_jobs/opportunity_management_jobs.py

```diff
@@ -314,69 +315,70 @@ def _generate_opportunities_related_to_topic(
+            set(stories_dict.keys()).intersection(story_ids))
+        exp_ids: List[str] = list(itertools.chain.from_iterable(
+            stories_dict[story_id].story_contents.get_all_linked_exp_ids()
+            for story_id in existing_story_ids))
+        existing_exp_ids = set(exps_dict.keys()).intersection(exp_ids)
+
+        missing_story_ids = set(story_ids).difference(existing_story_ids)
+        missing_exp_ids = set(exp_ids).difference(existing_exp_ids)
+        if len(missing_exp_ids) > 0 or len(missing_story_ids) > 0:
+            return result.Err(
+                'Failed to regenerate opportunities for topic id: %s, '
+                'missing_exp_with_ids: %s, missing_story_with_ids: %s' % (
+                    topic.id,
+                    list(missing_exp_ids),
+                    list(missing_story_ids)))
+
+        exploration_opportunity_summary_list = []
+        stories = [
+            stories_dict[story_id] for story_id in existing_story_ids
+        ]
+
+        exploration_opportunity_summary_model_list = []
+        with datastore_services.get_ndb_context():
+            for story in stories:
+                for exp_id in story.story_contents.get_all_linked_exp_ids():
+                    try:
                         exploration_opportunity_summary_list.append(
                             opportunity_services.create_exp_opportunity_summary(
                                 topic, story, exps_dict[exp_id]))
-
-                for opportunity in exploration_opportunity_summary_list:
-                    model = (
-                        opportunity_models.ExplorationOpportunitySummaryModel(
-                            id=opportunity.id,
-                            topic_id=opportunity.topic_id,
-                            topic_name=opportunity.topic_name,
-                            story_id=opportunity.story_id,
-                            story_title=opportunity.story_title,
-                            chapter_title=opportunity.chapter_title,
-                            content_count=opportunity.content_count,
-                            incomplete_translation_language_codes=(
-                                opportunity
-                                .incomplete_translation_language_codes
-                            ),
-                            translation_counts=opportunity.translation_counts,
-                            language_codes_needing_voice_artists=(
-                                opportunity
-                                .language_codes_needing_voice_artists
-                            ),
-                            language_codes_with_assigned_voice_artists=(
-                                opportunity
-                                .language_codes_with_assigned_voice_artists
-                            )
+                    except Exception as e:
+                        logging.exception(e)
+                        return result.Err((exp_id, e))
```

Comment:
Can we also mention in the error string that the exception occurred while processing the opportunities for the topic? Otherwise, the output is a little confusing since each job run output corresponds to one topic.

Reply:
I had done this to include exp_id in the error other than that it's the same as earlier. I would highly suggest restructuring the one-off job such that we can get error reports for all the opportunity models at once, currently it stops once there's one exploration of a topic that fails.

Reply:
Sounds good.